libcnb/src/layer/public_interface.rs

```diff
@@ -0,0 +1,192 @@
+    fn update(
+        &self,
+        context: &BuildContext<Self::Buildpack>,
+        layer_data: &LayerData<Self::Metadata>,
+    ) -> Result<LayerResult<Self::Metadata>, <Self::Buildpack as Buildpack>::Error> {
+        LayerResultBuilder::new(layer_data.content_metadata.metadata.clone()).build()
+    }
+
+    /// This method will be called by libcnb when parsing of the layer's metadata into
+    /// [`Self::Metadata`] failed. The result of this method call determines which strategy libcnb
+    /// should use to continue.
+    ///
+    /// The simplest strategy, [`MetadataMigration::RecreateLayer`] will delete the layer and
+    /// recreate it from scratch. This is also the default implementation.
+    ///
+    /// In some cases, a layer might be able to migrate metadata from an other version to a new
+    /// structure. For this, [`MetadataMigration::ReplaceMetadata`] can be used. Implementations can
+    /// use the raw TOML metadata passed as `metadata` to the method and the contents of the layer
+    /// to construct a new value for [`Self::Metadata`].
+    ///
+    /// # Implementation Requirements
+    /// Implementations **MUST** be read-only. They **MUST NOT** modify the file-system or write
+    /// anything to stdout/stdout or any other stream.
+    fn migrate_incompatible_metadata(
+        &self,
+        context: &BuildContext<Self::Buildpack>,
+        metadata: &GenericMetadata,
+    ) -> Result<MetadataMigration<Self::Metadata>, <Self::Buildpack as Buildpack>::Error> {
+        Ok(MetadataMigration::RecreateLayer)
+    }
+}
+
+/// The result of a [`Layer::migrate_incompatible_metadata`] call.
+pub enum MetadataMigration<M> {
+    /// The layer should be recreated entirely.
+    RecreateLayer,
+    /// The layer's metadata should be replaced by this new value.
+    ReplaceMetadata(M),
+}
+
+/// Information about an existing CNB layer.
+pub struct LayerData<M> {
+    pub name: String,
+    /// The layer's path, should not be modified outside of a [`Layer`] implementation.
+    pub path: PathBuf,
+    pub env: LayerEnv,
+    pub content_metadata: LayerContentMetadata<M>,
+}
+
+/// The result of a function that processes layer data.
+///
+/// Essentially, this carries additional metadata about a layer this later persisted according
+/// to the CNB spec by libcnb.
+pub struct LayerResult<M> {
+    pub metadata: M,
+    pub env: Option<LayerEnv>,
+}
+
+/// A builder that simplifies the creation of [`LayerResult`] values.
+pub struct LayerResultBuilder<M> {
+    metadata: M,
+    env: Option<LayerEnv>,
+}
+
+impl<M> LayerResultBuilder<M> {
+    pub fn new(metadata: M) -> Self {
+        Self {
+            metadata,
+            env: None,
+        }
+    }
+
+    pub fn env(mut self, layer_env: LayerEnv) -> Self {
+        self.env = Some(layer_env);
+        self
+    }
+
+    pub fn build<E>(self) -> Result<LayerResult<M>, E> {
+        Ok(LayerResult {
```

Comment:
You got it right, I made it return `Result` to avoid (IMO) noisy wrapping into `Ok` everywhere. It's meant as a small QOL enhancement. We might want to add a `build_unwrapped` or similar to at least allow it be used without a `Result` if necessary (i.e. when the buildpack code does `map` and existing result so they don't have to resort to `and_then`).

> our other builders (eg DetectResultBuilder and BuildResultBuilder) don't return a Result.

I planned to introduce the same pattern for those, just not in this PR to keep it shoter. ;-) 

> where it looks like the Ok() is missing given usage of ? or Err() previously

I do return `Result` values quite often as the last expression without using `Err` or `?`. Let's assume `LayerResultBuilder` would be actually fallible, what would you do here? The code below seems silly to me:

```rust
Ok(LayerResultBuilder::new(GenericMetadata::default()).build()?)
```